File: src/main/java/tk/deltarays/discordconsole/DiscordSocket.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//

package tk.deltarays.discordconsole;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.net.URI;
import java.util.*;

public class DiscordSocket extends WebSocketClient {
    public Timer timer;
    String lastS;
    Boolean isInvalid = false;
    String botId = "";
    DiscordConsole main;
    String sessionId = "";
    Boolean resume;
    Boolean ackReceived = true;
    private Boolean isOpening = true;
    private Boolean isConnected = false;

    public DiscordSocket(URI serverUri, DiscordConsole main, String sessionId) {
        super(serverUri);
        this.main = main;
        this.sessionId = sessionId;
        this.resume = true;
    }

    public DiscordSocket(URI serverUri, DiscordConsole main) {
        super(serverUri);
        this.main = main;
        this.resume = false;
    }

    public Boolean isOpening() {
        return this.isOpening;
    }

    public void onOpen(ServerHandshake serverHandshake) {
        if (this.main.getConfig().getBoolean("debug")) {
            this.main.getLogger().info("[WebSocket] Socket opened!");
        }

        this.isOpening = false;
    }

    public void onMessage(String msg) {
        JSONParser parser = new JSONParser();
        JSONObject parsed = new JSONObject();

        try {
            parsed = (JSONObject) parser.parse(msg);
        } catch (ParseException e) {
            this.main.getLogger().severe("[Discord WebSocket] Error encountered in parsing json payload");
            e.printStackTrace();
        }

        long op = (Long) parsed.get("op");
        String s = String.valueOf(parsed.get("s"));
        if (s != null) {
            this.lastS = s;
        }

        if (this.main.getConfig().getString("botToken") == null || !Objects.requireNonNull(this.main.getConfig().getString("botToken")).matches("^(?i)[a-z0-9.\\-_]{32,100}$")) {
            this.isInvalid = true;
            Bukkit.getScheduler().runTask(this.main, () -> this.main.getLogger().severe(ChatColor.DARK_RED + "An invalid bot token was provided! Go to the plugins folder, DiscordConsole, config.yml to modify the bot token"));
        }
        System.out.println(op);
        String content;
        if (op == 10L) {
            JSONObject d = (JSONObject) parsed.get("d");
            long heartbeat_interval = (Long) d.get("heartbeat_interval");
            int hb_i = Math.toIntExact(Math.round((float) heartbeat_interval));
            Bukkit.getServer().getConsoleSender().sendMessage(ConRef.getPlPrefix() + ChatColor.YELLOW + " Connecting to the discord bot...");
            JSONObject response;
            if (!this.resume) {
                content = this.main.getConfig().getString("botStatus");
                ArrayList statuses = new ArrayList(Arrays.asList("online", "dnd", "invisible", "idle"));
                if (!statuses.contains(content)) {
                    content = "online";
                }

                JSONObject game = new JSONObject();
                game.put("name", ConRef.replaceExpressions(this.main.getConfig().getString("botStatusText"), true));
                game.put("type", 0);
                JSONObject presResp = new JSONObject();
                if (this.main.getConfig().getString("botStatusText") != null && !this.main.getConfig().getString("botStatusText").isEmpty()) {
                    presResp.put("game", game);
                }

                presResp.put("status", content);
                JSONObject dProps = new JSONObject();
                dProps.put("$os", "linux");
                dProps.put("$browser", "DiscordSocket");
                dProps.put("$device", "DiscordConsole");
                JSONObject dResp = new JSONObject();
                dResp.put("token", this.main.getConfig().getString("botToken"));
                dResp.put("properties", dProps);
                dResp.put("presence", presResp);
                response = new JSONObject();
                response.put("op", 2);
                response.put("d", dResp);
            } else {
                JSONObject dResp = new JSONObject();
                dResp.put("token", this.main.getConfig().getString("botToken"));
                dResp.put("session_id", this.sessionId);
                dResp.put("seq", this.lastS);
                response = new JSONObject();
                response.put("op", 6);
                response.put("d", dResp);
            }

            this.send(response.toJSONString());
            this.timer = new Timer();
            this.timer.scheduleAtFixedRate(new TimerTask() {
                public void run() {
                    try {
                        if (!DiscordSocket.this.ackReceived) {
                            if (DiscordSocket.this.main.getConfig().getBoolean("debug")) {
                                DiscordSocket.this.main.getLogger().info("[Discord WebSocket] Hearbeat ack not received! Terminating the connection and reconnecting!");
                                DiscordSocket.this.main.socketConnect(DiscordSocket.this.lastS);
                                DiscordSocket.this.close(1002);
                                DiscordSocket.this.timer.cancel();
                                if (DiscordSocket.this.timer != null) {
                                    DiscordSocket.this.timer.purge();
                                }
                            }
                        } else if (!DiscordSocket.this.isInvalid && !DiscordSocket.this.isClosed() && !DiscordSocket.this.isClosing()) {
                            if (DiscordSocket.this.main.getConfig().isSet("debug") && DiscordSocket.this.main.getConfig().getBoolean("debug")) {
                                DiscordSocket.this.main.getLogger().info("[Discord WebSocket] Heartbeat sent");
                            }

                            String s = String.format("{\"op\": 1, \"d\": %s}", DiscordSocket.this.lastS != null ? DiscordSocket.this.lastS : "null");
                            DiscordSocket.this.send(s);
                            DiscordSocket.this.ackReceived = false;
                        } else {
                            DiscordSocket.this.timer.cancel();
                            DiscordSocket.this.timer.purge();
                        }
                    } catch (Exception e) {
                        DiscordSocket.this.main.getLogger().severe("[Discord WebSocket] Error in sending heartbeat!");
                        e.printStackTrace();
                        if (DiscordSocket.this.timer != null) {
                            DiscordSocket.this.timer.cancel();
                            DiscordSocket.this.timer.purge();
                        }
                    }

                }
            }, hb_i - 1, hb_i);
        } else if (op == 9L) {
            final boolean d = (Boolean) parsed.get("d");
            Timer t = new Timer();
            t.schedule(new TimerTask() {
                public void run() {
                    if (!d) {
                        DiscordSocket.this.main.socketConnect();
                    } else {
                        DiscordSocket.this.main.socketConnect(DiscordSocket.this.sessionId);
                    }

                }
            }, (long) (Math.random() * 4000.0D) + 1000L);
        } else if (op == 11L) {
            this.ackReceived = true;
        } else if (op == 0L) {
            String t = (String) parsed.get("t");
            JSONObject d = (JSONObject) parsed.get("d");
            if (t.contains("READY")) {
                this.isConnected = true;
                Bukkit.getServer().getConsoleSender().sendMessage(ConRef.getPlPrefix() + ChatColor.GREEN + " Successfully connected to the bot!");
                this.sessionId = (String) parsed.get("session_id");
                JSONObject user = (JSONObject) d.get("user");
                this.botId = (String) user.get("id");
            } else if (t.contains("RESUME")) {
                this.isConnected = true;
                this.resume = false;
            } else if (t.contains("MESSAGE_CREATE") && this.main.getConfig().getBoolean("consoleCommandsEnabled")) {
                String channelId = (String) d.get("channel_id");
                if (this.main.workingChannels.containsKey(channelId)) {
                    JSONObject author = (JSONObject) d.get("author");
                    String senderId = (String) author.get("id");
                    if (!senderId.equalsIgnoreCase(this.botId)) {
                        content = ((String) d.get("content")).replaceAll("\\\\\"", "\"").replaceAll("\\\\\\\\", "\\\\");

                        try {
                            String finalContent = content;
                            Bukkit.getScheduler().runTask(this.main, () -> Bukkit.getServer().dispatchCommand(Bukkit.getServer().getConsoleSender(), finalContent));
                        } catch (Exception e) {
                            this.main.getLogger().severe("[Discord WebSocket] Error in running command from console!");
                            e.printStackTrace();
                        }
                    }
                }
            }
        }

    }

    public Boolean isConnected() {
        return this.isConnected;
    }

    public void onClose(int code, String reason, boolean remote) {
        this.isConnected = false;
        if (this.timer != null) {
            this.timer.cancel();
            this.timer.purge();
        }

        if (this.main.getConfig().getBoolean("debug")) {
            this.main.getLogger().info(String.format("[Discord WebSocket] Closed! Code: %s Reason: %s", code, reason));
        }

        if (code == 4004) {
            this.isInvalid = true;
            this.main.getLogger().severe("An invalid discord bot token was provided!");
        } else if (code != 1001 && code != 1006) {
            if (code != 1002) {
                this.main.getLogger().severe(String.format("[Discord WebSocket] WebSocket closed!\nCode: %s\nReason: %s", code, reason));

                try {
                    this.main.socketConnect(this.sessionId);
                } catch (Exception e) {
                    this.main.getLogger().severe("[Discord WebSocket] Failure to reconnect!");
                    e.printStackTrace();
                }
            }
        } else {
            String token = this.main.getConfig().getString("botToken");
            if (token != null && !token.matches("^(?i)[a-z0-9.\\-_]{32,100}$")) {
                this.main.getLogger().severe("Invalid token provided");
            }

            this.main.getLogger().info("[Discord WebSocket] Got disconnected, reconnecting...");
            this.main.socketConnect(this.sessionId);
        }

    }

    public void onError(Exception e) {
        this.isConnected = false;
        this.main.getLogger().severe("[Discord Websocket] Error encountered!");
        e.printStackTrace();
    }
}

File: src/main/kotlin/me/deltarays/discordconsole/DiscordSocket.kt
package me.deltarays.discordconsole


import com.google.gson.JsonObject
import com.google.gson.JsonParser
import kotlinx.coroutines.*
import okhttp3.OkHttpClient
import okhttp3.Request
import org.java_websocket.client.WebSocketClient
import org.java_websocket.handshake.ServerHandshake
import java.net.URI
import kotlin.math.ceil

class DiscordSocket(uri: URI) : WebSocketClient(uri) {
    private lateinit var plugin: DiscordConsole
    var lastS: String? = null
    private var sessionId: String? = null
    private var isConnected = false
    private lateinit var botId: String
    private var isInvalid = false
    private lateinit var job: Job

    companion object {
        fun getWSUrl(): String? {
            val client = OkHttpClient()
            val request = try {
                (Request.Builder()).url("https://discordapp.com/api/v8/gateway?v=8&encoding=json").build()
            } catch (e: Exception) {
                return null
            }
            val response = try {
                client.newCall(request).execute()
            } catch (e: Exception) {
                return null
            }
            val parser = JsonParser()
            val obj = try {
                parser.parse(response.body()?.string())
            } catch (e: Exception) {
                return null
            }
            return obj.asJsonObject.get("url").asString
        }
    }

    fun setHandlingPlugin(pl: DiscordConsole) {
        this.plugin = pl
    }

    override fun onOpen(handshakedata: ServerHandshake?) {
        if (plugin.config.getBoolean("debug", false))
            Utils.logColored(
                plugin.getConfigManager().getPrefix(), "&a[Socket] Socket opened!", LogLevel.DEBUG
            )
    }

    private val parser = JsonParser()
    private var ackReceived = true
    override fun onMessage(message: String?) {
        Utils.logColored(plugin.getConfigManager().getPrefix(), "[Socket] Message received!\n$message", LogLevel.DEBUG)
        val payload: JsonObject = try {
            parser.parse(message).asJsonObject
        } catch (e: Exception) {
            if (plugin.config.getBoolean("debug", false)) {
                Utils.logColored(
                    plugin.getConfigManager().getPrefix(), "&4[WebSocket] Error!", LogLevel.SEVERE
                )
                e.printStackTrace()
            } else Utils.logColored(
                plugin.getConfigManager().getPrefix(),
                "&c[Discord Connection] Error!\nMessage: &f" + e.message + "\n&cSet debug to true in the config to find out more!",
                LogLevel.SEVERE
            )
            return
        }
        val op: Int = payload.get("op").asInt
        val unparsedS = payload.get("s")
        if (!unparsedS.isJsonNull) this.lastS = unparsedS.asString
        if (op == 10) {
            if (this.sessionId == null) {
                val heartbeatInterval = payload.get("d").asJsonObject.get("heartbeat_interval").asLong
                job = GlobalScope.launch(Dispatchers.IO) {
                    while (true) {
                        delay(heartbeatInterval)
                        if (isInvalid) {
                            this.cancel("invalid")
                        }
                        if (!ackReceived) {

                            close(3333, "No Heartbeat ACK received!")
                        }
                        val obj = JsonObject().apply {
                            addProperty("op", 1)
                            addProperty("d", lastS)
                        }
                        ackReceived = false
                        send(obj.toString())
                    }
                }
                val obj = JsonObject().apply {
                    addProperty("op", 2)
                    add("d", JsonObject().apply {
                        addProperty("token", plugin.getConfigManager().getBotToken())
                        addProperty("intents", 1 shl 9)
                        add("properties", JsonObject().apply {
                            addProperty("\$os", "unknown")
                            addProperty("\$browser", "DiscordSocket")
                            addProperty("\$device", "DiscordConsole")
                        })
                    })
                }
                send(obj.toString())
            } else {
                val obj = JsonObject().apply {
                    addProperty("op", 6)
                    add("d", JsonObject().apply {
                        addProperty("token", plugin.getConfigManager().getBotToken())
                        addProperty("session_id", sessionId)
                        addProperty("seq", lastS)
                    })
                }
                send(obj.toString())
            }
        } else if (op == 0) {
            when (payload.get("t").asString) {
                "READY" -> handleReady(payload)
                "RESUMED" -> handleResumed()
                "MESSAGE_CREATE" -> handleMessage()
            }
        } else if (op == 9) {
            Utils.logColored(
                plugin.getConfigManager().getPrefix(),
                "&e[Discord Connection] Unable to resume from when the bot was last connected! Creating a new connection...",
                LogLevel.WARNING
            )
            this.sessionId = null
            this.lastS = null
            GlobalScope.launch(Dispatchers.Default) {
                delay(ceil(Math.random() * 5).toLong())
                reconnect()
            }
        } else if (op == 7) {
            this.close(3333, "Discord asked me to reconnect")
        } else if (op == 11) {
            this.ackReceived = true;
        }

    }

    private fun handleReady(payload: JsonObject) {
        isConnected = true
        val d = payload.get("d").asJsonObject
        this.sessionId = d.get("session_id").asString
        this.botId = d.get("user").asJsonObject.get("id").asString
        Utils.logColored(
            plugin.getConfigManager().getPrefix(),
            "&a[Discord Connection] Successfully connected!",
            LogLevel.INFO
        )
    }

    private fun handleResumed() {
        Utils.logColored(
            plugin.getConfigManager().getPrefix(),
            "&a[Discord Connection] Bot has successfully resumed from where it last disconnected!",
            LogLevel.INFO
        )
    }

    private fun handleMessage() {

    }

    override fun onClose(code: Int, reason: String?, remote: Boolean) {
        this.isConnected = false
        job.cancel("closed")
        if (plugin.config.getBoolean("debug", false)) {
            Utils.logColored(
                plugin.getConfigManager().getPrefix(), String.format(
                    "[WebSocket] Closed!\nRemote: %s\nCode: %s\nReason:%s",
                    remote,
                    code,
                    reason
                ), LogLevel.DEBUG
            )
        }
        when (code) {
            4004 -> {
                Utils.logColored(
                    plugin.getConfigManager().getPrefix(), "&cAn invalid bot token was provided!", LogLevel.SEVERE
                )
                this.isInvalid = true
                return
            }
            4000 -> {
                Utils.logColored(
                    plugin.getConfigManager().getPrefix(),
                    "&e[Discord Connection] Discord encountered an unknown error! Reconnecting...",
                    LogLevel.WARNING
                )
                this.reconnect()
            }
            4001, 4002, 4003, 4005, 4007, 4012, 4013, 4014 -> {
                Utils.logColored(
                    plugin.getConfigManager().getPrefix(),
                    "&4[Discord Connection] The plugin's author made an oopsie! Contact him on discord (https://discord.gg/WSaWztJ) and tell him that you encountered an issue with" +
                            String.format(
                                "code: %s and reason: %s", code, reason
                            ),
                    LogLevel.SEVERE
                )
                this.isInvalid = true
            }
            4008 -> {
                Utils.logColored(
                    plugin.getConfigManager().getPrefix(),
                    "&e[Discord Connection] The bot was disconnected for doing something too fast! Reconnecting...",
                    LogLevel.WARNING
                )
                this.reconnect()
            }
            4009 -> {
                Utils.logColored(
                    plugin.getConfigManager().getPrefix(),
                    "&e[Discord Connection] The bot timed out! Reconnecting...",
                    LogLevel.INFO
                )
                this.reconnect()
            }
            3333 -> {
                Utils.logColored(
                    plugin.getConfigManager().getPrefix(),
                    "&e[Discord Connection] The bot was disconnected! Reconnecting...",
                    LogLevel.INFO
                )
                this.reconnect()
            }
        }

    }

    override fun onError(e: Exception?) {
        if (plugin.config.getBoolean("debug", false)) {
            Utils.logColored(
                plugin.getConfigManager().getPrefix(), "&4[WebSocket] Error!", LogLevel.SEVERE
            )
            e?.printStackTrace()
        } else Utils.logColored(
            plugin.getConfigManager().getPrefix(),
            "&c[Discord Connection] Error!\nMessage: &f" + e?.message + "\n&cSet debug to true in the config to find out more!",
            LogLevel.SEVERE
        )
    }
}